### DISCUSSION TITLE
First experience with Stride coming from Unity

### DISCUSSION THREAD

**Discussion by manio143** [stride3d/stride#1761] [General]
Given the recent influx of interest in our engine from devs who previously worked on Unity we want to be able to further provide people with opinions on why they should (or shouldn't) choose the Stride engine.

Please share your experiences below.
Try to avoid asking specific coding questions, preferably open a new discussion if not asked before. General questions are welcome.

If you haven't already, see the blog post;
https://www.stride3d.net/blog/embracing-open-source-stride-as-an-alternative-to-unity/

**Comment by iDrmzIt** [comment]
Hello all, I've been modding and developing with the Unity engine since 2016, starting with approximately five years of modding experience, to coding my own project in Unity for the last two.

Starting with Stride is definitely a different experience, a lot of quality of life that Unity provides Stride currently does not. There are many good libraries made by people to extend Stride's capabilities, which I cannot speak to the quality of just yet, but they seem to cover a decent amount of features you may miss from Unity.

This engine is much more hands on than Unity. Just from some preliminary testing, there are no guardrails on what you can do with the engine. You don't like something? Change it. Maybe even post it to make the engine better. I may only have a year and a half of .NET experience, but I believe if you are a curious person and want that hands-on boots-on-the-ground feeling of being able to truly pioneer your own features, Stride Engine may be there for you.

At least so far from what I've done, I would recommend at least trying it if you like C# and .NET programming, especially if you come from Unity, the component system is nearly identical. 

Overall, I recommend trying it. It will have growing pains and a learning curve, but the customization is in your hands.

**Comment by JoeAbiN** [comment]
I've installed Xenko a few years ago but only opened it like twice because I was already using Unity, but now I'm interested in Stride.

What I like about Stride is that it is a purely C# and open-source engine and its interface, component system and API feel like Unity but with less convolution since it's basically a pure C# project, so I recommend this engine for people who use Unity and/or are invested in C# and .NET.

There are concerns though, like the lack of important features like an animation state machine implementation. But I understand that this is an open-source project with people working on it in their free time as opposed to the big companies that have many more people and teams. That being said, I really like the active and welcoming community and I hope to see it grow.

**Comment by Pooffick** [comment]
Hi, this is copy of my discord message:

I’ve studied all official tutorials on youtube and some docs. So the main obstacles for me now are:
1) poor fbx import (single model/animation per file)
2) poor animations (no “animator”, no masks/layers, no event triggers by frame)
3) poor 2d animations (I can’t say “frame 0 to 12 is idle, frame 13 to 24 is walk, etc.” or combine different sprites into one animation)
4) there is no timeline analog (I need to make cutscenes in code)

Problems 3 and 4 are not as important as problems 1 and 2.

In any case, the engine seems very promising to me, I will try to use it in some upcoming jams.

**Reply by FOSSSoftwareUser** [reply]
2D animation is a lot easier to do in code, it is just a few lines. Look up XNA tutorials, they should have code for doing 3D animations too.

**Reply by Pooffick** [reply]
@FOSSSoftwareUser, yes, that's what I'll do. As I said above, this problem is not very important.

**Comment by ExtraMedicated** [comment]
I just started playing around with this today, and I'm trying to port some scripts over from a Unity project. A lot of things seem fairly simple to rewrite, but there are some functions I haven't found a direct replacement for yet, such as directly setting the world rotation of a transform component.

My Unity projects also relied on Final IK to get my non-humanoid characters to look at and pick up things. I don't think the engine itself really needs built-in IK solvers as long as there are plugins or something available, but I haven't found any yet.

The graphics out of the box look like they might be better than Unity's. I noticed the neat little glare effect of the light hitting my untextured model.

**Comment by ACCIAI0** [comment]
I find the lack of a way to create custom editors and editor extensions as a possible brick wall for devs in the future of the engine. Another thing that should be integrated is the equivalent of the [SerializeField] attribute for private properties and fields, since not always we want those variables to be publicly accessible and editable.

**Reply by makotech222** [reply]
> equivalent of the [SerializeField]

I think you want [DataMember]

**Reply by ly3027929699** [reply]
Yeah, I need it too.I like the Extensions in unity editor, such as Odin Inspector which can Draw inspector with some custom method.

**Reply by Eideren** [reply]
> the equivalent of the [SerializeField] attribute for private properties and fields

Init properties cannot be set by other scripts, but you can set them through the editor, something like so:
``public float MyEditorValue { get; init; }``
It still gives other scripts the possibility to read the value contrary to a private field, but I don't believe that's as big an issue ?

**Reply by IXLLEGACYIXL** [reply]
or you structure your project in subprojects and have an internal setter os in your main project no setter is visible

**Comment by FOSSSoftwareUser** [comment]
This will be the first Unity 3D replacement I will try because of Vulcan support, C# code, Linux platform support and it is free and open software. I would like to be able to use another C# editor so I don't have to sign a Microsoft account.

**Comment by ly3027929699** [comment]
I like the Extensions in unity editor, such as Odin Inspector which can Draw inspector with some custom method.I can even execute code in the editor by clicking a custom button, which is awesome. I don't know of some way to achieve this.
I particularly like this pure C# coding method. This is an absolute advantage. I can reference any package and use the latest features of .net. I am looking forward to .net8.0.

**Reply by FlameskyDexive** [reply]
Editor extentions is so important for client developer and designer, hope it will coming soon..

**Reply by BenjaBobs** [reply]
Threads to keep an eye on:
- https://github.com/stride3d/stride/issues/1120
- https://github.com/stride3d/stride/issues/1629
- https://github.com/stride3d/stride/discussions/1031

**Reply by WazWaz** [reply]
It's a bad look for an engine to not to implement the UI of the Editor in that engine. People extending the editor don't want to learn whatever UI system you choose _in addition_ to that supported by the engine.

**Comment by dlfily** [comment]
It's very similar to Unity, I really like it, and I quickly got used to changes,I hope to add scenes like Unity that can see the status during runtime。
I don't know if it has been implemented

**Comment by WazWaz** [comment]
Day 1:

Pros:
- Best C# experience. Better than Unity, far better than Godot.
- Open Source - being able to drill down into the source helps when documentation isn't clear.
- License (of course).

Cons:
- Asset workflow feels awkward: change .blend, manually export to fbx, manually click Update Assets, find .blend file in Resources to edit again even though that is completely invisible in the editor.
- I'm daunted by the lack of a Visual Shader Editor, but I'm liking the look of SDSL so far (nothing written yet though).
- No anti-aliasing on SDF fonts - is that right? Doesn't that basically come for free?

Some little things that are annoying but that only a new user would realise:
- Game Settings is fine as an Asset, but it's rather disconcerting when you don't know that and find Edit/Settings. Needs a menu link (Project/Settings).
- Default 1280x720 that templates default to (at least RPG Template) looks terrible (tiny window, upscaled if fullscreened), and is horribly slow with default settings (I'm not sure what, I changed so much). Eventually I managed to see it run at 144Hz 4K just fine. This is terrible a "advertisement" for the platform. The results should be AWESOME right out of the box, because people will assume the samples show off the best case, not the worst.

**Comment by AzureBerylBlue** [comment]
What is the Stride equivalent of Unity's humanoid animations? (The feature in Unity where you can set the rig on the models and animations to humanoid and everything just works, even when their skeletons don't match)

**Comment by sandsalamand** [comment]
Pros:

- Pure C#
- Free and Open Source
- No magic methods
- Overall API design seems a bit better than Unity
- Async Support

Cons:

- No live scene view while the game is running
- Lack of editor extensibility / plugins
- Lack of Linux support for Game Studio. This wouldn't be a problem if it looked like something that could be changed in the near future, but from reading the discussions, it seems like WPF and MSBuild are deeply integrated into Stride, so it will be hard/impossible to change it any time soon.

**Comment by SVNMLR** [comment]
Here are some features, possible improvements & thoughts I have for now using Stride the past Week.

# Launcher
- Include release notes from all builds, not only from major Updates (maybe with a button next to the Buttons for selecting and deleting the stride version?)

# General
- Editor UI is not properly saving its last states (Tab and window sizes and positions)
- Settings: Provide changes to some colors in the Sceneeditor (e.g. the selected Objects wireframe / the Grid / Pivot Colors) -> See Unity
- Provide more availbale Editor Keybindings in Settings
- Modelimport: All submeshes and objects seperate, skinned meshes, no merging -> prevent Hierarchy
- Bool to be able disable entities like in Unity
- Make private variables available in Inspector/Properties grid
- Visual scripting
- Effect graph out of the Box
- Material/Shader graph out of the Box
- Lanscape Tools out of the Box
- Wind Generators e.g. for Tree & Grass movement
- Animator / visual Animation Tools out of the Box


# Main Scene
- If rotating around the Object: If Alt + Mouse are beeing pressed while focused it is possible to accitentally move the Object if the Mousecursor ist over the Pivot. Either ignore it or duplicate the Object, like in Unreal ?
- Collider selectable by mouseclick
- Selecting multiple Objects by dragging a rectangle with the left Mouse button is not available (Cursor gets just hided and nothing is happening)


- View: While Moving around (Right Mousebutton + WSDA) -> Scrollwheel changes the MoveSpeed -> RightMouseButton+Scrollwheel (already working on that for myself for a PR)
- View: Mouse-Panning is not available if Alt is pressed (already working on that for myself for a PR)
- View: Slightly ease out camera movement (already working on that for myself for a PR)
- The toggle for snapping is not highlighted enough (Already in PR #1807)

# Entity Properties
- A Prefab Override need to be more present
- Change icon to Quick change on hover a Property with Mouse from a cross to horizontal arrows
- Adding a Component like a Rigidbody, automatically add a Collider with the size ob the overall bounds of the Object
- Have to use Tab twice to get to the next inputfield and be able to edit it. Once only selects the field itself

# Output / Debug
- Colored Symbols for Toggle Buttons -> Info/Warning/Error/Fatal (Already in PR #1807)

**Comment by AndrewRChapman** [comment]
I scanned some of the forum entries here and get the impression, most are interesting in the coding aspects of Unity/Stride.

Whilst I used to code, I became 'code blind' and therefore only use Unity/Unreal/Stride etc etc as an Artist, creating scene and environment props, along with a handful of games using nothing more than the engine collision aspects.

My first impression of Stride, from that perspective, is that it has a long way to go with Mesh Importing, Materials and other such features.

I have brought in some of the variety of models/meshes/texture maps etc, that I sell on Unity/Unreal/SL etc etc.... into Stride and achieved some reasonable results given the importer and material handling. This shows that all elements of my creations will port into Stride, except Materials, building the Materials within the Stride Editor, seems a bit long winded and therefore the importers need more work.

The other danger with a public contributing system, is the Blender Syndrome..... where things get put in place by committee after committee and can complicate creativity, which the sole reason behind such an endeavour in the first place. Blender has become a 'Can Do It All' piece of kit.... often overwhelming the User, as opposed to bringing the productivity aspects to the finger tips, become buried under reams of nested menus.

I could not find access to a Development Roadmap anywhere, hopefully there is one and I have simply overlooked it.

I prefer to be able to create my own editor window, where I can have regular used functions to hand, in windows of my making, in positions and scales of my choice. In both Unreal and Unity, I can drag and drop 'windows' / panels into most locations and tabs.... but without coding, I cannot create my own Favourites Menus..... A big mistake in my opinion.

Similarly, in other editors, there is a tendency to throw all assets into one big folder, whereas I have always put Meshes, Materials, Textures into their own bespoke folders. A quirky feature in Unreal, as a Publisher, they will only accept a submission if every Mesh has a SM_ prefix, every texture a Tex_ and a Material a Mat_ prefix, Blueprints must also have a prefix.... which is all manual and tedious when your project includes 100's or 1,000's of components. All that despite the engine knowing the attributes of each component and displays that fact when you hover over it.... superfluous and tedious. Unity on the other hand insists on every component used in the Scene, have a Prefab created with Zeroed Attributes.....

In my opinion, a good team that creates a VR engine, should be well balanced between Artist User/Developer and Code/Script Developer.

I wish the team a fair wind. I would also add, all the Products on Unity Asset Store are compatible with Stride, except the Scripted Items, which is Security Camera systems, those bind with the core scripts of the Unity Engine. Meshes are FBX or occasionally OBJ, and textures/maps are PNG. The same products are available on Unreal, they are formulated to make them a Unreal Asset format only. If anyone wants to know the workflow for importing a Unity asset into Stride, even without a Unity Engine, please feel free to message me.

**Reply by NicusorN5** [reply]
As a development roadmap, have you looked into https://opencollective.com/stride3d/projects ?

**Reply by AndrewRChapman** [reply]
Hey NicusorN5, thanks for the link..... whilst useful, it is perhaps shown as a wish list? I'm not trying to play down the enormous tasks ahead and the contributions of so many, but as a Projects manager for years, I can foresee the advancements forging ahead in some areas, and falling behind in others.... just like Blender did for many years.....

I am sure everyone is aware, that everything must be done in a sequence (cart and horse) for it to work. Not easy when you rely so heavily on a community spirit. In a perfect world, such a balance would be there from day one. I will certainly keep my eye on Stride and would like to see an opening page of a Game, say Made With Pride, in Stride.

**Reply by NicusorN5** [reply]
>it is perhaps shown as a wish list?

You could say it is a kind of wishlist. In a way, I think that what roadmaps are in a way.

>  that everything must be done in a sequence

Well, I think that's just the nature of FOSS software. Features are constantly worked on at the same time by developers and contributors.

**Reply by tebjan** [reply]
To understand how Stride operates, there is info in the beginning of this blog post: https://www.stride3d.net/blog/announcing-stride-4-2-in-dotnet-8

**Comment by AndrewRChapman** [comment]
> To understand how Stride operates, there is info in the beginning of this blog post: https://www.stride3d.net/blog/announcing-stride-4-2-in-dotnet-8
Hello tebjan, I looked at the post, which exposes a lot of the hard work that goes into the creation/development of this engine. Thanks for the link.
